allennlp/data/dataset_readers/nlvr2_reader.py

```diff
@@ -106,17 +106,21 @@ def __init__(
         text_path_prefix: str,
         visual_path_prefix: str,
         topk_images: int = -1,
+        tokenizer: Tokenizer = None,
+        token_indexers: Dict[str, TokenIndexer] = None,
```

Comment:
If the only correct choices here are of the `PretrainedTransformer*` variety, and they have to match, then I think this should only take `model_name: str = "bert-base-uncased"` parameter.

Reply:
Yeah, I thought about that, though I think there's only one place that we actually require this, and it's in logic that only exists because of a small research project that Sanjay did.  I'm not sure we need to keep that logic in this version of the class.

To be more concrete: a `PretrainedTransformerTokenizer` is only required if you use `mask_prepositions_verbs`.  Instead of restricting the types to only be `PretrainedTransformer*` objects here, I'd instead either:
1. Remove the `{drop,mask}_prepositions_verbs` functionality from this reference implementation as not needed
2. Raise an error if you don't have the right tokenizer when using `mask_prepositions_verbs`
3. Come up with a workaround for how to mask prepositions and verbs when you're not using a pretrained transformer (though this probably doesn't make much sense).

I put those in my order of preference.  I'm not feeling too well today, so I'm not going to be working on this, but I'll do option 1 the next time I write code, unless anyone has objections.

Reply:
I agree that this functionality might not be needed in this reference implementation, but that's orthogonal to how the tokenizer and indexer are specified, isn't it?

Reply:
Yes, but it's that functionality that is the only thing placing constraints on this.

Oh, unless you're saying that the model requires the use of a pretrained transformer?  Well, vilbert does, but you could use this dataset reader with other models that don't make that assumption.  There's nothing in the reader itself that requires that use, so I was keeping as much flexibility in the types as possible.